### ISSUE TITLE
AVX-512 compile error on macOS Clang

### ISSUE THREAD

**Issue by tritao** [BLAKE3-team/BLAKE3#79]
### Describe the bug

Clean clone of Wasmer fails to compile on MacOS due to AVX-512 error when compiling assembly code.

The actual compilation command that fails (full compilation log below):

`"cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-std=c11" "-o" "wasmer/target/debug/build/blake3-da163d92031ce8e8/out/c/blake3_avx512_x86-64_unix.o" "-c" "c/blake3_avx512_x86-64_unix.S"`

```sh
cargo:warning=c/blake3_avx512_x86-64_unix.S:2315:9: error: instruction requires: AVX-512 ISA AVX-512 VL ISA
cargo:warning=        vpxord xmm3, xmm3, xmm0
```

Compilation works if  `-march=skylake-avx512` is added to command above, the error happens because Clang by default does not enable AVX-512 intrinsics since they are not supported by my CPU.

### Additional context

Full compilation log:

```sh
error: failed to run custom build command for `blake3 v0.3.1`

Caused by:
  process didn't exit successfully: `wasmer/target/debug/build/blake3-32d62a0f85ffa978/build-script-build` (exit code: 1)
--- stdout
cargo:rerun-if-env-changed=CARGO_FEATURE_PURE
cargo:rerun-if-env-changed=CARGO_FEATURE_PURE
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
CC_x86_64-apple-darwin = None
CC_x86_64_apple_darwin = None
HOST_CC = None
CC = None
CFLAGS_x86_64-apple-darwin = None
CFLAGS_x86_64_apple_darwin = None
HOST_CFLAGS = None
CFLAGS = None
CRATE_CC_NO_DEFAULTS = None
CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2,sse3,ssse3")
CC_x86_64-apple-darwin = None
CC_x86_64_apple_darwin = None
HOST_CC = None
CC = None
CFLAGS_x86_64-apple-darwin = None
CFLAGS_x86_64_apple_darwin = None
HOST_CFLAGS = None
CFLAGS = None
CRATE_CC_NO_DEFAULTS = None
CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2,sse3,ssse3")
cargo:rerun-if-env-changed=CARGO_FEATURE_PREFER_INTRINSICS
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
CC_x86_64-apple-darwin = None
CC_x86_64_apple_darwin = None
HOST_CC = None
CC = None
CFLAGS_x86_64-apple-darwin = None
CFLAGS_x86_64_apple_darwin = None
HOST_CFLAGS = None
CFLAGS = None
CRATE_CC_NO_DEFAULTS = None
CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2,sse3,ssse3")
CC_x86_64-apple-darwin = None
CC_x86_64_apple_darwin = None
HOST_CC = None
CC = None
CFLAGS_x86_64-apple-darwin = None
CFLAGS_x86_64_apple_darwin = None
HOST_CFLAGS = None
CFLAGS = None
CRATE_CC_NO_DEFAULTS = None
CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2,sse3,ssse3")
cargo:rerun-if-env-changed=CARGO_FEATURE_PURE
cargo:rerun-if-env-changed=CARGO_FEATURE_PREFER_INTRINSICS
cargo:rustc-cfg=blake3_sse41_ffi
cargo:rustc-cfg=blake3_avx2_ffi
cargo:rustc-cfg=blake3_avx512_ffi
TARGET = Some("x86_64-apple-darwin")
OPT_LEVEL = Some("0")
HOST = Some("x86_64-apple-darwin")
CC_x86_64-apple-darwin = None
CC_x86_64_apple_darwin = None
HOST_CC = None
CC = None
CFLAGS_x86_64-apple-darwin = None
CFLAGS_x86_64_apple_darwin = None
HOST_CFLAGS = None
CFLAGS = None
CRATE_CC_NO_DEFAULTS = None
DEBUG = Some("true")
CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2,sse3,ssse3")
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-std=c11" "-o" "wasmer/target/debug/build/blake3-da163d92031ce8e8/out/c/blake3_sse41_x86-64_unix.o" "-c" "c/blake3_sse41_x86-64_unix.S"
exit code: 0
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-std=c11" "-o" "wasmer/target/debug/build/blake3-da163d92031ce8e8/out/c/blake3_avx2_x86-64_unix.o" "-c" "c/blake3_avx2_x86-64_unix.S"
exit code: 0
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-std=c11" "-o" "wasmer/target/debug/build/blake3-da163d92031ce8e8/out/c/blake3_avx512_x86-64_unix.o" "-c" "c/blake3_avx512_x86-64_unix.S"
cargo:warning=c/blake3_avx512_x86-64_unix.S:28:9: error: instruction requires: AVX-512 ISA
cargo:warning=        kmovw k1, r9d
cargo:warning=        ^
cargo:warning=c/blake3_avx512_x86-64_unix.S:38:9: error: instruction requires: AVX-512 ISA AVX-512 VL ISA
cargo:warning=        vpcmpltud k2, ymm2, ymm0
cargo:warning=        ^
```

```sh
rustc 1.42.0 (b8cedc004 2020-03-09) | x86_64
```

Output of `sysctl -a | grep machdep.cpu.features`:

```sh
machdep.cpu.features: FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE SSE3 PCLMULQDQ DTES64 MON DSCPL VMX EST TM2 SSSE3 FMA CX16 TPR PDCM SSE4.1 SSE4.2 x2APIC MOVBE POPCNT AES PCID XSAVE OSXSAVE SEGLIM64 TSCTMR AVX1.0 RDRAND F16C
```

To note, no support for AVX-512.

Related issue from Wasmer: https://github.com/wasmerio/wasmer/issues/1373

**Comment by oconnor663** [comment]
We do have macOS machines in our Github Actions CI runs, and I can confirm on [at least this run](https://github.com/BLAKE3-team/BLAKE3/runs/577154487?check_suite_focus=true) that they don't support AVX-512. So maybe this problem is specific to a certain compiler version? I will add some printing to the CI run to show exactly what compiler version is in use, and I'll see if I can repro this on a local macOS machine. `clang version 9.0.1` does not repro on my Linux machine.

> Compilation works if -march=skylake-avx512 is added to command above

This suggests a likely workaround will be to add `-mavx512f -mavx512vl` to that invocation of the compiler, as we do with the intrinsics-based implementation. (Runtime CPU feature detection will prevent us from actually executing this code on a machine that doesn't support the instructions.) With GCC it seems like

**Comment by oconnor663** [comment]
Seems like it might be related to these:

https://bugs.llvm.org/show_bug.cgi?id=39875
https://bugs.llvm.org/show_bug.cgi?id=36202

Those links report that the problem may have been fixed with Clang 7.0. It might make sense to work around older Clang nonetheless?

**Comment by tritao** [comment]
> Seems like it might be related to these:
> 
> https://bugs.llvm.org/show_bug.cgi?id=39875
> https://bugs.llvm.org/show_bug.cgi?id=36202
> 
> Those links report that the problem may have been fixed with Clang 7.0. It might make sense to work around older Clang nonetheless?

Saw that as well when initially searching for the issue, but disregarded it because in my case Clang is pretty recent.

```
clang --version
Apple LLVM version 10.0.0 (clang-1000.11.45.2)
Target: x86_64-apple-darwin17.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by oconnor663** [comment]
Huh. Can you confirm that `-march=skylake-avx512` does work for you, as suggested above? And just to be sure, could you clarify how you're setting the flags?

**Comment by oconnor663** [comment]
I'm not able to repro the failure on the macOS Catalina laptop I have here. This command succeeds:

`"cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-std=c11" "-o" "target/debug/build/blake3-787565f4b6ccf4dc/out/c/blake3_avx512_x86-64_unix.o" "-c" "c/blake3_avx512_x86-64_unix.S"`

Version info:

```
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin19.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by oconnor663** [comment]
I just added a "print compiler version" step to each CI build. You can see it in [this run](https://github.com/BLAKE3-team/BLAKE3/runs/580040993). It looks like all the macOS CI machines run Clang 11. Is this a 10 vs 11 difference? (This is using the Apple version numbers, which I believe are different from the version numbers we see on Linux.)

**Comment by sneves** [comment]
[Apple LLVM 10 corresponds to upstream LLVM 6.0.1](https://en.wikipedia.org/wiki/Xcode#Toolchain_versions) so the bug Jack pointed out makes sense.

**Comment by sneves** [comment]
Taking a LLVM 6.0.1 build, we can get it to work with `./clang+llvm-6.0.1-x86_64-linux-gnu-ubuntu-16.04/bin/clang -mavx512f -mavx512vl -mavx512dq blake3_avx512_x86-64_unix.S`

**Comment by sneves** [comment]
7ef795d62effe2f59e7aff6d4bbba81e398d23e3 makes it not need `-mavx512dq` anymore.

**Comment by tritao** [comment]
I tried to build with that fix on my system but still same result.

> Taking a LLVM 6.0.1 build, we can get it to work with `./clang+llvm-6.0.1-x86_64-linux-gnu-ubuntu-16.04/bin/clang -mavx512f -mavx512vl -mavx512dq blake3_avx512_x86-64_unix.S`

Can also confirm this works.

Should those arguments be added to `build.rs` in the `build_x86_asm` section?

**Comment by sneves** [comment]
Something like 

```diff
diff --git a/build.rs b/build.rs
index c2572aa..e3e7122 100644
--- a/build.rs
+++ b/build.rs
@@ -166,6 +166,8 @@ fn build_x86_asm() {
         // explicit workaround for macOS also.
         build.file("c/blake3_sse41_x86-64_unix.S");
         build.file("c/blake3_avx2_x86-64_unix.S");
+        build.flag("-mavx512f");
+        build.flag("-mavx512vl");
         build.file("c/blake3_avx512_x86-64_unix.S");
     }
     build.compile("blake3_ffi");
```

should work, I guess, but I'd rather not add it unconditionally. You might find yourself with a fairly old gcc version but a binutils that supports AVX-512, and compilation would fail then. So this should only be added if it can be established that the compiler is clang <= 6.0.1 with the integrated assembler.

**Comment by oconnor663** [comment]
Yes I need to make some build.rs changes. I've also been interested in refactoring things a little bit, so that for example even if you can't handle the AVX-512 assembly, you still get the SSE4.1 and AVX2 assembly, instead of automatically falling back to intrinsics-only.

> You might find yourself with a fairly old gcc version but a binutils that supports AVX-512, and compilation would fail then

The current behavior is to check whether the compiler supports `-mavx512` or whatever, and to not even try to compile/assemble any AVX-512 code if it doesn't. (Currently that means falling back to Rust SSE4.1 and AVX2 intrinsics only, but the refactoring I'm talking about doing above will change that to falling back to SSE4.1 and AVX2 assembly instead on x86_64.) So in the case you're describing, the build will still work, but the user will miss out on the AVX-512 implementation, even though in theory they could've successfully assembled it. Do you think it's worth trying to improve that case (binutils supports AVX-512 even though the compiler doesn't), or is that obsolete enough that it's not worth bothering?

**Comment by sneves** [comment]
Is it not possible to disable the feature on compilation failure, and not rely on whether `-mxxxx` flags exist?

Either way, that case is fairly unlikely, so I don't have strong opinions one way or another on how to handle it.

**Comment by oconnor663** [comment]
That's an option, yes. We could try to build the default way, catch any failures that result, and then try to build again with some different approach. The reason I haven't done that so far, though, is that I'm worried about masking unexpected errors. Like if I go into the AVX-512 assembly and insert some dumb typo, I really want that to break the build. I don't want my bug to cause everyone to silently fall back to a different implementation.

**Comment by oconnor663** [comment]
@tritao could you check whether this branch builds on your system (`cargo test` etc): https://github.com/BLAKE3-team/BLAKE3/tree/buildrs

**Comment by tritao** [comment]
Sure, just tried it and everything worked, all tests pass, including AVX-512 ones. 👍

**Comment by oconnor663** [comment]
Awesome, I'll land it later tonight. The AVX-512 tests will appear to pass, even if your machine doesn't support AVX-512, because they'll short-circuit without testing anything :p (Unfortunately `cargo test` doesn't have a way to represent skipping a test at runtime.)

**Comment by oconnor663** [comment]
Just published v0.3.2 with this fix.
